### PR TITLE
Add support for XML encoding

### LINKED ISSUE #22: Add support for XML format?
Certain APIs don't support JSON (e.g. openstreetmaps). Wouldn't it be a good idea to be able to select between a few possible encodings? Maybe by adding a new field to napping.Session which would default to json. That way it's backward compatible but can be extended to use xml or whatnot.

If you want, I could try and prepare a pull request for this, if you're interested.

### PR DESCRIPTION
**Pull request by gdm85**
Fixes #22
Use URL field when constructing test server URL in tests
Re-organize context-aware log output to use '%+v' instead of JSON (former util.go)

Tests are still missing, please advise with early feedback :)

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -12,18 +12,25 @@ requests (cookies, auth, proxies).
...
 	"net/http"
 	"net/url"
+	"path/filepath"
+	"runtime"
 	"strings"
 	"time"
 )
 
-import ()
+const (
```
Comment (jmcvetta) [inline session.go]: It would be nice to have some commentary here briefly describing how these contants will be used.

File: session.go [hunk 2]
```diff
@@ -319,3 +333,48 @@ func (s *Session) log(args ...interface{}) {
 		log.Println(args...)
 	}
 }
+
```
Comment (jmcvetta) [inline session.go]: Needs a doc comment

File: session.go [hunk 3]
```diff
@@ -319,3 +333,48 @@ func (s *Session) log(args ...interface{}) {
...
+	case JSON:
+		s.logWithContext(response.RawText())
+		return
+	case XML:
+		s.logWithContext(response.RawText())
+		return
+	}
+	panic("invalid encoding")
+}
+
```
Comment (jmcvetta) [inline session.go]: Needs a doc comment

File: session.go [hunk 4]
```diff
@@ -319,3 +333,48 @@ func (s *Session) log(args ...interface{}) {
...
+func (s *Session) logWithContext(v interface{}) {
```
Comment (jmcvetta) [inline session.go]: Needs a doc comment

**Comment by AaronO** [comment]
@gdm85 It's not a bad idea, but I don't think we should `panic` in a library that kind of goes against go's whole error handling philosophy. I think you should pass errors back to the caller rather than `panic`.

@jmcvetta What do you think ?

File: session.go (hunk 2 above)
Comment (gdm85) [inline session.go]: I thought only exported functions need that?

**Comment by gdm85** [comment]
@AaronO that panic is there because you're typing random numbers in the encoding field instead of using the enum constants, not a runtime problem but something to catch during development :)

File: session.go (hunk 1 above)
Comment (gdm85) [inline session.go]: Sure. I thought these could be used as enum constants e.g.
```
    s := napping.Session{
       Encoding: napping.XML,
    }
```

By default Encoding is 0, which corresponds to the JSON enum constant.

**Comment by jmcvetta** [comment]
Only exported functions *need* a doc comment.  But it is nice to have them even on unexported functions.  :)

**Comment by gdm85** [comment]
Sure, I'll add more.

I am also thinking of adding some tests.

**Comment by jmcvetta** [comment]
Yes, please add tests to cover the new functionality.

As for the panic - I agree this is not a recoverable error.  However the panic in question would be the only panic statement in Napping, and I don't really like that.  Probably better to pass an error back to the caller.  

However I'm very open to commentary from other users.

**Comment by gdm85** [comment]
I think it's a valid usage of ``panic()`` because - alternatively - one could refactor the code to have a private field (``encoding``) and only set it via two factory methods (``NewJSONSession`` and ``NewXMLSession``).

Instead we use a more declarative syntax, but also make sure that error cannot be ignored (a returned ``error`` can be ignored, we know that right?) during development.

Plus, the panic will hit any test written by the users of the library, making it obvious what they are doing wrong (same way as if you provided factory methods as explained above).

**Comment by gdm85** [comment]
@jmcvetta I have added some comments, please let me know if I shall add more and be more clear/verbose :)

I am now going to add another commit for the tests

**Comment by gdm85** [comment]
With my last commit I am refactoring the encoding/(un)marshalling into an interface, in an attempt to split it away from other logic and make it more easily testable.

**Comment by jmcvetta** [comment]
Looking good so far.

**Comment by jmcvetta** [comment]
Sorry for the delay - had to do a manual merge on this one.  

Thanks for contributing!

**Comment by gdm85** [comment]
@jmcvetta on the contrary, sorry for not completing this! I think it's still missing some proper tests for the XML side of things...

**Comment by jmcvetta** [comment]
Please add some tests, and I will never them too!  :)

-- sent from my robot

**Comment by jmcvetta** [comment]
Hey, I just had to [revert](https://github.com/jmcvetta/napping/commit/fafb50eafd72f471f589f067016513ddab5d2968) this merge.

  I didn't notice when I accepted it - it changed the Napping API in a way that was incompatible with existing consumers of the package.  That caused problems for people:
* https://github.com/jmcvetta/neoism/issues/80
* https://github.com/jmcvetta/neoism/pull/81

I'm still quite open to adding your functionality.  Please adjust your changes so they do not break backward compatibility, add some tests, and submit a new PR.

**Pull request merged**